hbase_storage/src/main/java/com/splicemachine/storage/HScan.java

```diff
@@ -43,6 +48,57 @@ public boolean isDescendingScan(){
         return scan.isReversed();
     }
 
+    /**
+     * Take a list of [startKey, stopKey) rowkey pairs, where the stopKey is excluded,
+     * convert it to the corresponding {@link MultiRowRangeFilter} and attach
+     * it to the {@link Scan} held in this {@link HScan} as a {@link Filter}.
+     *
+     * @param rowkeyPairs the list of [startKey, stopKey) pairs to convert.
+     *
+     * @return if this {@link HScan} currently has no filter, build a new
+     * {@link MultiRowRangeFilter} out of {@code rowkeyPairs} and attach it
+     * as a filter.  If this {@link HScan} already has a {@link MultiRowRangeFilter},
+     * replace it with a new MultiRowRangeFilter built from {@code rowkeyPairs}.
+     * If {@code rowkeyPairs} has no elements, do not build a filter.
+     *
+     * @throws IOException
+     *
+     * @notes A possible future enhancement is, instead of replacing an old
+     * MultiRowRangeFilter with a new one, build the intersection of the
+     * old filter and the new one.
+     *
+     * @see     Scan
+     */
+    @Override
+    public void addRowkeyRangesFilter(List<Pair<byte[],byte[]>> rowkeyPairs) throws IOException {
+        if (rowkeyPairs == null || rowkeyPairs.size() < 1) {
+            return;
+        }
+        Filter currentFilter = scan.getFilter();
+        if (currentFilter != null && !(currentFilter instanceof MultiRowRangeFilter))
+            return;
+
+        List<MultiRowRangeFilter.RowRange> ranges = new ArrayList<>(rowkeyPairs.size());
+        byte[] startKey;
+        byte[] stopKey;
+        for (Pair<byte[],byte[]> startStopKey: rowkeyPairs) {
+            startKey=startStopKey.getFirst();
+            stopKey=startStopKey.getSecond();
+            MultiRowRangeFilter.RowRange rr =
+            new MultiRowRangeFilter.RowRange(startKey, true,
+                                             stopKey, false);
```

Comment:
The startKey/stopKey pair isn't really a point range.  We always add a trailing byte of 01xB on to the stop key so that we read all the matching keys.  01xB is higher than anything that could match.  For example, if we have primary key (a,b), a start key of -127 might represent primary key value a = 1, with b not specified.  But, we also want to read, for example, a row with the key a=1, b=1, which is an internal rowkey of :
startRow = {byte[3]@11822} 
 0 = -127
 1 = 0
 2 = -127

The stop row that we build : (-127, 1) covers all possible values of b, because byte 1 in a valid rowkey will always be less than 1.  So, if we did mark the stopKey as inclusive, it would never match any valid primary key values, and have no effect.  Also, stopKeys are usually exclusive in HBase.  Perhaps this is why splice also builds exclusive stop keys.

In the call to buildStartAndStopKeys, you can also see that the start key is marked GE ( >= ), and the stop key is marked GT ( > ) : 
                  buildStartAndStopKeys(startKeyDVDs, ScanController.GE,
                                        stopKeyDVDs, null,
                                        ScanController.GT,

Not sure why we don't call it LT ( < ).

Reply:
I see, thanks for the explanation!